kamon-core/src/main/scala/kamon/module/ModuleRegistry.scala

```diff
@@ -269,63 +266,39 @@ class ModuleRegistry(classLoading: ClassLoading, configuration: Configuration, c
-            val name = moduleClazz.getName()
-            val description = "Module detected from the legacy kamon.reporters configuration."
-
-            Module.Settings("unknown", name, description, moduleClazz, inferredModuleKind, true)
-          }
-
-          if(emitConfigurationWarnings) {
-            moduleSettings.failed.foreach(t => _logger.error(s"Failed to load legacy reporter module [${moduleClass}]", t))
-          }
-
-          moduleSettings
-        })
-        .filter(_.isSuccess)
-        .map(_.get)
-
-
-      val (repeatedLegacyModules, uniqueLegacyModules) = legacyModuleSettings
-        .partition(lm => moduleSettings.find(_.clazz.getName == lm.clazz.getName).nonEmpty)
-
-      if(emitConfigurationWarnings) {
-        repeatedLegacyModules.foreach(m =>
-          _logger.warn(s"Module [${m.name}] is configured twice, please remove it from the deprecated kamon.reporters setting."))
-
-        uniqueLegacyModules.foreach(m =>
-          _logger.warn(s"Module [${m.name}] is configured in the deprecated kamon.reporters setting, please consider moving it to kamon.modules."))
-      }
-
-      moduleSettings ++ uniqueLegacyModules
-
-    } else moduleSettings
   }
 
   /**
     * Creates a module from the provided settings.
     */
-  private def createModule(settings: Module.Settings): Option[Entry[Module]] = {
-    val moduleInstance = classLoading.createInstance[Module](settings.clazz, Nil)
-    val moduleEntry = moduleInstance.map(instance => createEntry(settings, false, instance))
+  private def createModule(settings: Module.Settings, programmaticallyAdded: Boolean): Option[Entry[Module]] = {
+    val moduleEC = createExecutor(settings)
+    val moduleEntry = for {
+      moduleConfigPath  <- Try(settings.configPath.get)
+      moduleConfig      <- Try(Kamon.config().getConfig(moduleConfigPath))
+      factory           <- classLoading.createInstance[ModuleFactory](settings.factory.get, Nil)
+    } yield {
+      val instance = factory.create(ModuleFactory.Settings(moduleConfig, moduleConfigPath, moduleEC))
+      Entry(settings.name, moduleEC, programmaticallyAdded, settings, instance)
+    }
 
-    moduleEntry.failed.foreach(t => _logger.warn(s"Failed to create instance of module [${settings.name}]", t))
+    moduleEntry.failed.foreach(t => {
+      moduleEC.shutdown()
+      _logger.warn(s"Failed to create instance of module [${settings.name}]", t)
+    })
     moduleEntry.toOption
   }
 
-  private def createEntry(settings: Module.Settings, programmaticallyAdded: Boolean, instance: Module): Entry[Module] = {
+  private def createExecutor(settings: Module.Settings): ExecutionContextExecutorService = {
     val executor = Executors.newSingleThreadExecutor(threadFactory(settings.name))
-    Entry(settings.name, ExecutionContext.fromExecutorService(executor), programmaticallyAdded, settings, instance)
+
+    // Scheduling any task on the executor ensures that the underlying Thread is created and that the JVM will stay
```

Comment:
Why is this desired? Couldn't this cause the JVM to hang on exit if there is an issue? I think that could be extremely confusing fo rusers

Reply:
The idea is to force people to shutdown the modules so that they have a final chance to report any accumulated data, we even have [some tests](https://github.com/kamon-io/Kamon/blob/master/kamon-core-tests/src/test/scala/kamon/KamonLifecycleSpec.scala#L16-L32) covering this behavior and https://github.com/kamon-io/Kamon/issues/533 that are sort of related to this.

If an user sees that while testing their JVM doesn't shutdown then their are much more likely to take this into account and properly call Kamon.stopModules() before shutting down the app.

Reply:
I like the idea of the users have a final chance to report theirs accumutated data but I not sure if return a `ExecutionContextExecutorService` is a good idea thinking in the java users, maybe a simple `java.util.concurrent.ExecutorService` is enough.

Reply:
This should be probably documented somewhere. We've just came into this issue of hanging application because of Kamon modules. I'd expect somewhere in installation guide that user should handle stopping of modules.